### DISCUSSION TITLE
Has MediatR been used incorrectly in this project.

### DISCUSSION THREAD

**Discussion by mike-azility** [LuckyPennySoftware/MediatR#930] [Q&A] [answered]
I have been given a project with no contact possible with the previous developer.   I have not used MediatR or the mediator pattern before but I have the feeling it has been done incorrectly in this project. 
I have calls like this in the Controllers.
`var supplier = _mediator.Send(new GetSupplierDetails.Query(Convert.ToInt32(id), entityId)).Result;` 
Where `GetSupplierDetails`  class has `public class QueryHandler : DataHandler<Query, SupplierModel>`   and Task Handle that runs some SQL queries.  
This means `GetSupplierDetails` is hardcoded into the calling Controllers.   From my limited understanding of MediatR the whole idea is for that not to be the case? 
Instead the calls should look more like
`var supplier = _mediator.Send(new SupplierModel(Convert.ToInt32(id), entityId)).Result;`  where `SupplierModel` is just a class that holds the data?  (syntax may not be correct). 
Are they wrong or am I missing something?

**Comment by jbogard** [comment, marked as answer]
Oh that is kinda weird, I've not had the query DTO be an inner class of the handler itself. Sometimes folks make the handler and DTOs be inner classes of the same empty parent class, but a namespace will do the same trick. I do place these classes adjacent to each other on disk though, since queries and handlers are naturally cohesive (they change with each other).

The bigger error here is not using `await`. The `Send` method is async and should be awaited. Calling `Result` can result in deadlocks.
